### PR TITLE
[inductor] print triton float64 constants correctly

### LINKED ISSUE #134720: index out of bounds: 0 <= tmp4 < libdevice.trunc(0.120000000000000*(ks0.to(tl.float64))).to(tl.int32)
### 🐛 Describe the bug

Meta internal report: https://fburl.com/workplace/jqcuiynj

Repro:
```
import logging
from dataclasses import dataclass
from datetime import datetime
from typing import Dict

import torch
import torch._inductor.config as inductor_config
import torch.nn as nn

logger = logging.getLogger(__name__)
TIME_FORMAT_STR: str = "%b_%d_%H_%M_%S"


@dataclass
class BenchmarkConfig:
    batch_size: int = 1024
    enable_bf16: bool = True
    enable_pt2: bool = False
    enable_unit_pt2_wrap: bool = True
    device = "cuda"
    dtype = torch.bfloat16
    d_model = 384
    top_k_value = 0.12
    max_length_dict = {"event_a": 1000, "event_b": 200, "event_c": 200}


class SimpleSequenceSummarizationBlock(nn.Module):

    # @lint-ignore FIXIT [TooManyArgsInFunction]
    def __init__(
        self,
        benchmark_config: BenchmarkConfig,
    ) -> None:
        super().__init__()

        self.d_model = benchmark_config.d_model

        self.attn_norm = nn.LayerNorm(self.d_model)

        self.self_gating = nn.Linear(self.d_model, self.d_model)
        self.top_k_value = benchmark_config.top_k_value
        self.router = nn.Linear(self.d_model, 1)

    def forward(
        self,
        x: torch.Tensor,
    ) -> torch.Tensor:

        bs, seq_len, d = x.shape
        # select top k tokens
        routing_w = self.router(x).squeeze(-1)  # shape [B, Seq]
        top_k_w, top_k_indices = torch.topk(
            routing_w, int(self.top_k_value * seq_len), dim=1
        )
        # top_k sorting
        top_k_indices_in_order, ordering = torch.sort(top_k_indices, dim=1)
        top_k_w_in_order = top_k_w.gather(dim=1, index=ordering)

        x_compute_indices = top_k_indices_in_order.unsqueeze(-1).expand(-1, -1, d)
        x_compute_scale = top_k_w_in_order.unsqueeze(-1).expand(-1, -1, d)
        x_compute = x.gather(dim=1, index=x_compute_indices)

        attn_output = self.attn_norm(x_compute)

        x_compute = x_compute + attn_output * x_compute_scale
        x_full = x.clone()
        x_full.scatter_(1, x_compute_indices, x_compute)
        x = x_full

        # Self gating
        x = self.self_gating(x)

        return x


class SimpleSeqSummarizationLayer(torch.nn.Module):
    def __init__(
        self,
        benchmark_config: BenchmarkConfig,
    ) -> None:
        super().__init__()

        self.seq_summarization_modules: nn.ModuleDict = nn.ModuleDict()
        self.event_features_mapping = list(benchmark_config.max_length_dict.keys())
        for event_name in self.event_features_mapping:
            self.seq_summarization_modules[event_name] = (
                SimpleSequenceSummarizationBlock(
                    benchmark_config,
                )
            )

        self.enable_unit_pt2_wrap = benchmark_config.enable_unit_pt2_wrap
        if self.enable_unit_pt2_wrap:
            assert (
                not benchmark_config.enable_pt2
            ), "enable_unit_pt2_wrap and enable_pt2 cannot be both True"

    def forward(
        self,
        seq_embs: Dict[str, torch.Tensor],
    ) -> Dict[str, torch.Tensor]:
        # seq_summarizations = []
        seq_embs_after_summarizations = {}
        for event_name in self.event_features_mapping:
            if self.enable_unit_pt2_wrap:
                curr_seq_embs_after_summarizations = torch.compile(
                    self.seq_summarization_modules[event_name],
                )(
                    seq_embs[event_name],
                )
            else:
                curr_seq_embs_after_summarizations = self.seq_summarization_modules[
                    event_name
                ](
                    seq_embs[event_name],
                )
            seq_embs_after_summarizations[event_name] = (
                curr_seq_embs_after_summarizations
            )

        return seq_embs_after_summarizations


def create_model_input(benchmark_config: BenchmarkConfig):
    batch_size = benchmark_config.batch_size
    device = benchmark_config.device

    dtype = benchmark_config.dtype

    x = {}
    for event_name in benchmark_config.max_length_dict.keys():
        x[event_name] = torch.randn(
            (
                batch_size,
                benchmark_config.max_length_dict[event_name],
                benchmark_config.d_model,
            ),
            dtype=dtype,
            device=device,
            requires_grad=True,
        )
    return [x]


def run_first_model_once(model, input):
    pred = model(*input)
    pred = [i.sum() for i in pred.values()]
    pred = torch.stack(pred).sum()
    pred.sum().backward(retain_graph=True)


def single_run_benchmark():
    benchmark_config = BenchmarkConfig()
    model_input = create_model_input(benchmark_config)
    model = SimpleSeqSummarizationLayer(benchmark_config)

    model = model.to(benchmark_config.device)
    if benchmark_config.enable_bf16:
        model = model.to(dtype=torch.bfloat16)

    if benchmark_config.enable_pt2:
        inductor_config.decompose_mem_bound_mm = True
        inductor_config.trace.enabled = True
        model = torch.compile(model)
        model = model.to(benchmark_config.device)
        print("Start compiling model.")
        run_first_model_once(model, model_input)

    # trace
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        with_flops=True,
        schedule=torch.profiler.schedule(skip_first=1, wait=1, warmup=1, active=1),
        record_shapes=True,
        profile_memory=True,
        # with_stack=True,
    ) as profiler:
        for _ in range(10):
            run_first_model_once(model, model_input)
            profiler.step()

    return


def main() -> None:
    single_run_benchmark()
    print("done")


if __name__ == "__main__":
    main()
```

Leads to this assertion:
```
unknown:0: unknown: block: [4998,0,0], thread: [63,0,0] Assertion `index out of bounds: 0 <= tmp4 < libdevice.trunc(0.120000000000000*(ks0.to(tl.float64))).to(tl.int32)` failed.
```

### Versions

Recent main

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135260

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by masnesral** [comment]
Also tested repro provided in https://github.com/pytorch/pytorch/issues/134720

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -385,6 +385,12 @@ def _print_TruncToInt(self, expr):
             f"libdevice.trunc({self._print(expr.args[0])}).to({V.kernel.index_dtype})"
         )
 
+    def _print_Float(self, expr):
```
Comment (masnesral) [inline torch/_inductor/codegen/triton.py:388]: I admit I don't know if it's always valid to assume constants are float64. I'm operating on the assumption that any float literal originated from Python and is technically float64.

File: test/inductor/test_triton_kernels.py [hunk 2]
```diff
@@ -1545,6 +1545,17 @@ def f(x, y):
         x = torch.randn(4, device=GPU_TYPE)
         f(x, x)
 
+    @requires_gpu
+    def test_triton_kernel_float64_constant(self):
+        def f(x):
+            return x * (0.12 * x.shape[0])
+
+        x = torch.ones(200, device=GPU_TYPE, dtype=torch.float64)
```
Comment (isuruf) [inline test/inductor/test_triton_kernels.py:1553]: Can you parameterize the dtype here?

**Comment by ezyang** [comment]
Unfortunately, I can't tell if this correct. In particular, I don't know what the correct types of the Triton IR are supposed to be in this codegen case... Use of float64 here seems right, or at least, it is consistent with some of the other float codegen.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
I believe this is only for constants in indexing expressions, not all constants...

**Comment by lezcano** [comment]
This would turn every computation within indexing with floats into `fp64`, as this explicit casting makes any comptuation deriving from it to be performed in `fp64`, which may not be desirable.

https://github.com/triton-lang/triton/pull/4613 would fix this in the repro from https://github.com/pytorch/pytorch/issues/134720, just upcasting to `fp64` constants if they are going to be involved in a computation with an fp64 tensor.

File: test/inductor/test_triton_kernels.py (hunk 2 above)
Comment (masnesral) [reply test/inductor/test_triton_kernels.py:1553]: @isuruf , I don't think so? At least, there's no dtype that I know about. We just have a `sympy.core.numbers.Float` object that we're printing.

**Comment by masnesral** [comment]
> which may not be desirable

@lezcano , what problems would you anticipate? Of the people here, I certainly know the least about it. But if there's a float constant in an indexing expression, shouldn't that constant always be treated as fp64 to match eager semantics?

**Comment by lezcano** [comment]
GPUs have less dedicated silicon to compute fp64 than to compute fp32, even less so on consumer GPUs.

Regading eager semantics, if you do `torch.arange(8) * 3.0` you get an `fp32` not an `fp64`. With this PR everything will be computed on `fp64` tho

**Comment by masnesral** [comment]
@jansel WDYT about @lezcano's input here? How do you prefer to fix it? I've verified that https://github.com/triton-lang/triton/pull/4613 indeed fixes the original repro from https://github.com/pytorch/pytorch/issues/134720. I don't know know how big of a deal it is to get that Triton change. Do we update the pin or is it a cherry-pick situation?

**Comment by ezyang** [comment]
While @lezcano is right, I kind of think we should ship this change anyway. Mostly because I don't actually see anyway to recover Python-style float64 semantics when we move them to CUDA without doing it in float64. This is different from int32/int64, where we have a chance of optimizing it by value ranges analysis; I don't see anyway to see how to go from float64 to float32 and guarantee bit for bit equivalence in the end.

**Comment by masnesral** [comment]
Ok, landing this now given @ezyang's guidance (and because we have an internal usage waiting on a fix). If the discussion continues and we decide this is the wrong choice, we can always revert.

File: test/inductor/test_triton_kernels.py (hunk 2 above)
Comment (masnesral) [reply test/inductor/test_triton_kernels.py:1553]: Oh, misunderstood you comment :/

**Comment by jansel** [comment]
Yeah, indexing code is usually run in Python which is float64.  I also think this will be uncommon and not matter much for perf.  If it does we could optimize it.

File: test/inductor/test_triton_kernels.py (hunk 2 above)
Comment (masnesral) [reply test/inductor/test_triton_kernels.py:1559]: done

Comment (masnesral) [reply test/inductor/test_triton_kernels.py:1559]: ... except I got it wrong. omg. fix forthcoming

**Pull request closed without merging**